enrollment-server/src/main/java/com/wultra/app/enrollmentserver/impl/service/OtpService.java

```diff
@@ -129,34 +128,46 @@ public OtpVerifyResponse verifyOtpCode(String processId, String otpCode, OtpType
             throw new OnboardingProcessException();
         }
         OnboardingProcessEntity process = processOptional.get();
+
         Optional<OnboardingOtpEntity> otpOptional = onboardingOtpRepository.findLastOtp(processId, otpType);
         if (!otpOptional.isPresent()) {
             logger.warn("Onboarding OTP not found for process ID: {}", processId);
             throw new OnboardingProcessException();
         }
         OnboardingOtpEntity otp = otpOptional.get();
+
         // Verify OTP code
+        boolean expired = false;
         boolean verified = false;
-        int remainingAttempts = 0;
         int failedAttempts = onboardingOtpRepository.getFailedAttemptsByProcess(processId, otpType);
         int maxFailedAttempts = onboardingConfig.getOtpMaxFailedAttempts();
-        if (otp.getStatus() == OtpStatus.ACTIVE && failedAttempts < maxFailedAttempts) {
-            if (otp.getOtpCode().equals(otpCode)) {
-                verified = true;
-                otp.setStatus(OtpStatus.VERIFIED);
-                otp.setTimestampVerified(new Date());
-            } else {
-                otp.setFailedAttempts(otp.getFailedAttempts() + 1);
-                failedAttempts++;
-                if (failedAttempts >= maxFailedAttempts) {
-                    otp.setStatus(OtpStatus.FAILED);
-                    otp.setErrorDetail("maxFailedAttempts");
-                    // Onboarding process is failed, update it
-                    process.setStatus(OnboardingStatus.FAILED);
-                    process.setTimestampLastUpdated(new Date());
-                    process.setErrorDetail("maxFailedAttempts");
-                    process = onboardingProcessRepository.save(process);
-                }
+        if (OtpStatus.ACTIVE != otp.getStatus()) {
+            logger.warn("Unexpected not active {}, process ID: {}", otp, processId);
+        } else if (failedAttempts >= maxFailedAttempts) {
+            logger.warn("Unexpected OTP code verification when already exhausted max failed attempts, process ID: {}", processId);
+            process = failProcess(process, OnboardingOtpEntity.ERROR_MAX_FAILED_ATTEMPTS);
+        } else if (otp.hasExpired()) {
+            logger.info("Expired OTP code received, process ID: {}", processId);
+            expired = true;
+            otp.setStatus(OtpStatus.FAILED);
+            otp.setErrorDetail(OnboardingOtpEntity.ERROR_EXPIRED);
+            otp.setTimestampLastUpdated(new Date());
+            onboardingOtpRepository.save(otp);
+        } else if (otp.getOtpCode().equals(otpCode)) {
+            verified = true;
+            otp.setStatus(OtpStatus.VERIFIED);
+            otp.setTimestampVerified(new Date());
+            otp.setTimestampLastUpdated(new Date());
```

Comment:
Store `new Date()` in variable in the beginning, otherwise you can have different dates in subsequent lines. We should review all places with the `new Date()` code.

Reply:
Fixed